File: Managed/NextTurn.UE.Runtime/CoreUObject/SetProperty.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public sealed class SetProperty : Property, IProperty<ScriptSet>
    {
        internal SetProperty(IntPtr pointer) : base(pointer) { }

        public Property ItemProperty => Create<Property>(NativeMethods.GetItemProperty(this.pointer));

        public ScriptSet GetValue(Object @object, int index = 0) => this.GetValue<ScriptSet>(@object, index);

        public void SetValue(Object @object, ScriptSet value, int index = 0) => this.SetValue<ScriptSet>(@object, value, index);

        internal static new class NativeMethods
        {
            [ReadOffset]
            public static extern IntPtr GetItemProperty(IntPtr property);

            [ReadOffset]
            public static extern ScriptSet.Layout GetLayout(IntPtr property);
        }
    }
}

File: Source/NextTurnRuntime/Private/CoreUObject/SetProperty.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Accessor.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/UnrealType.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME SetProperty

TYPE_EXPORT_START

const FMemberSymbol Members[] =
{
	EXPORT_OFFSET_PUBLIC(FSetProperty, ItemProperty, ElementProp),
	EXPORT_OFFSET_PUBLIC(FSetProperty, Layout, SetLayout),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
